arrow/array/record.go

```diff
@@ -405,6 +406,84 @@ func (b *RecordBuilder) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+type iterReader struct {
+	refCount atomic.Int64
+
+	schema *arrow.Schema
+	cur    arrow.Record
+
+	next func() (arrow.Record, error, bool)
+	stop func()
+
+	err error
+}
+
+func (ir *iterReader) Schema() *arrow.Schema { return ir.schema }
+
+func (ir *iterReader) Retain() { ir.refCount.Add(1) }
+func (ir *iterReader) Release() {
+	debug.Assert(ir.refCount.Load() > 0, "too many releases")
```

Comment:
Technically this is a TOC/TOU error but it's a debug assertion so probably not a big deal